## Assignment 3: $k$ Nearest Neighbor


**Q1.**
1. What is the difference between regression and classification?
2. What is a confusion table? What does it help us understand about a model's performance?
3. What does the SSE quantify about a particular model?
4. What are overfitting and underfitting?
5. Why does splitting the data into training and testing sets, and choosing $k$ by evaluating accuracy or SSE on the test set, improve model performance?
6. With classification, we can report a class label as a prediction or a probability distribution over class labels. Please explain the strengths and weaknesses of each approach.

**Q2.** This is a case study on $k$ nearest neighbor classification, using the `land_mines.csv` data.

The data consists of a label, `mine_type`, taking integer values 1 to 5, and three properties of the mine, `voltage`, `height` and `soil`. We want to predict the kind of mine from data about it. Imagine working for the DOD or a humanitarian aid agency, trying to help people remove land mines more safely.

1. Load the data. Perform some EDA, summarizing the target label and the features.
2. Split the sample 50/50 into training and test/validation sets. (The smaller the data are, the more equal the split should be, in my experience: Otherwise, all of the members of one class end up in the training or test data, and the model falls apart.)
3. Build a $k$-NN classifier. Explain how you select $k$.
4. Print a confusion table for the optimal model, comparing predicted and actual class label on the test set. How accurate is it? Where is performance more or less accurate?
5. Notice that you can have a lot of accurate predictions for a given type of mine, but still make a lot of mistakes. Please explain how you'd advise someone to actually use this predictive model in practice, given the errors that it tends to make.

---

## *Phase 1: Solution without AI

Complete all questions in the following cells without generative AI. Then commit and push the
notebook before beginning Phase 2.

**Declaration:** I completed this version without generative AI.

**Student(s):** Jhanvi Patel

**Date:** 09/30/2026

In [ ]:
# Your Phase 1 solution to the problem goes here.


In [ ]:
Question 1
#1. Regression is used to predict a numerical value. Regression can be used to predict someone's house price or blood pressure. Classification is used when we want to predict a category or class, and it could be used to predict whether an email is spam or not spam. Regression predicts numbers and classification predicts categories.
#2. A confusion table compares the actual class of an observation to the class that the model predicted. It shows true positives, true negatives, false positives, and false negatives, which helps to understand not only how accurate the model is, but also what types of mistakes the model is making.
#3. SSE (Sum of Squared Errors), measures how far the model's predicted values are from the actual values. It adds up the squared differences between the actual and predicted values. A lower SSE means that the model's predictions are closer to the actual values, so the model fits the data better.
#4. Overfitting happens when a model is too complex and learns the training data too closely, including random noise. It can perform well on the training data but poorly on new data. Underfitting happens when a model is too simple and does not capture the important patterns in the data. An underfit model usually performs poorly on both the training and testing data.
#5. Splitting the data allows to train the model on one set of data and then test it on data that the model has not seen before. This helps to see how well the model will perform on new data instead of just seeing how well it memorizes the training data. By comparing accuracy or SSE for different values of k on the test set, we can choose the value of k that gives better performance and avoid overfitting.
#6. Example: The model could predict that a patient has a disease or does not have a disease. A class label does not tell us how confident the model is in its prediction. A probability distribution gives more information because it shows the probability of each possible class. For example, the model could say there is an 80% chance of disease and a 20% chance of no disease. This gives us more information about the model's confidence, but it can be more complicated to interpret and we may still need to choose a class based on the probabilities.

In [10]:
#Question 2.1
import pandas as pd
data = pd.read_csv('land_mines.csv')
print(data.head())
print(data.shape)
print(data.describe())
print(data['mine_type'].value_counts().sort_index())

#the The dataset has 338 observations and 4 variables
# mine_type is the target variable
# voltage, height, and soil are features used to predict the type of mine
# There are 5 different mine type
# Mine type 1: 71 observations
# Mine type 2: 70 observations
# Mine type 3: 66 observations
# Mine type 4: 66 observations
# Mine type 5: 65 observations
#The classes are around evenly distributed, there is not a major imbalance in mine_type.
#Features: the mean voltage is about 0.431, the mean height is about 0.509, and the mean soil value is about 0.504.

    voltage    height  soil  mine_type
0  0.338157  0.000000   0.0          1
1  0.320241  0.181818   0.0          1
2  0.287009  0.272727   0.0          1
3  0.256284  0.454545   0.0          1
4  0.262840  0.545455   0.0          1
(338, 4)
          voltage      height        soil   mine_type
count  338.000000  338.000000  338.000000  338.000000
mean     0.430634    0.508876    0.503550    2.952663
std      0.195819    0.306043    0.344244    1.419703
min      0.197734    0.000000    0.000000    1.000000
25%      0.309737    0.272727    0.200000    2.000000
50%      0.359516    0.545455    0.600000    3.000000
75%      0.482628    0.727273    0.800000    4.000000
max      0.999999    1.000000    1.000000    5.000000
mine_type
1    71
2    70
3    66
4    66
5    65
Name: count, dtype: int64


In [9]:
#Question 2.2

from sklearn.model_selection import train_test_split
X = data[["voltage", "height", "soil"]]
y = data["mine_type"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.5, random_state=42, stratify=y)

# The data is split into 50% test data and 50% training data.
# Stratification is used so that each type of mine is represented in both datasets

In [8]:
#Question 2.3

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

for k in range(1, 11):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    predictions = knn.predict(X_test)
    print(k, accuracy_score(y_test, predictions))

  #I used a k-nearest neighbors (k-NN) classifier to predict the type of mine based on its voltage, height, and soil measurements.
  #k-NN works by looking at the k observations that are closest to a new observation and using their mine types to make a prediction.
  #I tested different values of k and calculated the accuracy for each one.
  #I then chose the value of k with the highest accuracy because it performed the best at predicting the mine type.
  #The average validation accuracy for k=2 was 43.7%, so I used k=2 for the final model.


1 0.40236686390532544
2 0.4378698224852071
3 0.378698224852071
4 0.35502958579881655
5 0.3136094674556213
6 0.3254437869822485
7 0.3136094674556213
8 0.3136094674556213
9 0.34911242603550297
10 0.33727810650887574


In [7]:
#Question 2.4

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

knn = KNeighborsClassifier(n_neighbors=2)

knn.fit(X_train, y_train)

predictions = knn.predict(X_test)

print(confusion_matrix(y_test, predictions))

print(accuracy_score(y_test, predictions))

# The accuracy is 43.8%, so the model correctly predicted 43.8% of test observations
# In the matrix, the diagonal shows the correct predictions.
# Each row is a different mine type, and through the matrix, you can see that mine type 2 had the greatest prediction accuracy (32)
# The model had a difficult time predicting the accuracy of mine type 5 (0)
# Overall, the model was good at predicting mine type 1 and mine type 2, and worse at predicting the rest.

[[25  0  6  4  1]
 [ 0 32  0  3  0]
 [12  2  9  6  4]
 [12  5  8  8  0]
 [15  2 10  5  0]]
0.4378698224852071


In [6]:
#Question 2.5

# I would not recommend using this model by itself to decide whether a land mine is safe to remove. The model is only about 43.8% accurate overall, and it makes many mistakes between mine types 3, 4, and 5.
# A wrong classification could be very dangerous. I would use the model as an additional source of information that could help experts identify possible mine types.
# The prediction could be combined with other information and checked by trained professionals before a mine is removed.
# The model could be useful for narrowing down possible mines to remove, but it should not be treated as the sole reliable answer because of the number of incorrect predictions it makes.

## *Phase 2: AI-assisted Revision
Keep Phase 1 frozen.

- **Phase 1 commit SHA ID:** ce2c981a664bb46085074a441abbe6d13a3cc5f8
- **AI tool and model used:** Claude Sonnet 5.5

### *Prompts used

1. (For question 1)

ok so I have this assignment that I already did can u check if my answer is right and meets the requirements.

Q1.

1. What is the difference between regression and classification?
2. What is a confusion table? What does it help us understand about a model's performance?
3. What does the SSE quantify about a particular model?
4. What are overfitting and underfitting?
5. Why does splitting the data into training and testing sets, and choosing by evaluating accuracy or SSE on the test set, improve model performance?
6. With classification, we can report a class label as a prediction or a probability distribution over class labels. Please explain the strengths and weaknesses of each approach.

my answers:
Question 1
1. Regression is used to predict a numerical value. Regression can be used to predict someone's house price or blood pressure. Classification is used when we want to predict a category or class, and it could be used to predict whether an email is spam or not spam. Regression predicts numbers and classification predicts categories.
2. A confusion table compares the actual class of an observation to the class that the model predicted. It shows true positives, true negatives, false positives, and false negatives, which helps to understand not only how accurate the model is, but also what types of mistakes the model is making.
3. SSE (Sum of Squared Errors), measures how far the model's predicted values are from the actual values. It adds up the squared differences between the actual and predicted values. A lower SSE means that the model's predictions are closer to the actual values, so the model fits the data better.
4. Overfitting happens when a model is too complex and learns the training data too closely, including random noise. It can perform well on the training data but poorly on new data. Underfitting happens when a model is too simple and does not capture the important patterns in the data. An underfit model usually performs poorly on both the training and testing data.
5. Splitting the data allows to train the model on one set of data and then test it on data that the model has not seen before. This helps to see how well the model will perform on new data instead of just seeing how well it memorizes the training data. By comparing accuracy or SSE for different values of k on the test set, we can choose the value of k that gives better performance and avoid overfitting.
6. Example: The model could predict that a patient has a disease or does not have a disease. A class label does not tell us how confident the model is in its prediction. A probability distribution gives more information because it shows the probability of each possible class. For example, the model could say there is an 80% chance of disease and a 20% chance of no disease. This gives us more information about the model's confidence, but it can be more complicated to interpret and we may still need to choose a class based on the probabilities.


2. For question #2

ok this is the second question:

Q2. This is a case study on nearest neighbor classification, using the `land_mines.csv` data.
The data consists of a label, `mine_type`, taking integer values 1 to 5, and three properties of the mine, `voltage`, `height` and `soil`. We want to predict the kind of mine from data about it. Imagine working for the DOD or a humanitarian aid agency, trying to help people remove land mines more safely.

1. Load the data. Perform some EDA, summarizing the target label and the features.
2. Split the sample 50/50 into training and test/validation sets. (The smaller the data are, the more equal the split should be, in my experience: Otherwise, all of the members of one class end up in the training or test data, and the model falls apart.)
3. Build a -NN classifier. Explain how you select .
4. Print a confusion table for the optimal model, comparing predicted and actual class label on the test set. How accurate is it? Where is performance more or less accurate?
5. Notice that you can have a lot of accurate predictions for a given type of mine, but still make a lot of mistakes. Please explain how you'd advise someone to actually use this predictive model in practice, given the errors that it tends to make.


my answers:

#Question 2.1
import pandas as pd
data = pd.read_csv('land_mines.csv')
print(data.head())
print(data.shape)
print(data.describe())
print(data['mine_type'].value_counts().sort_index())

the The dataset has 338 observations and 4 variables
mine_type is the target variable
voltage, height, and soil are features used to predict the type of mine
There are 5 different mine type
Mine type 1: 71 observations
Mine type 2: 70 observations
Mine type 3: 66 observations
Mine type 4: 66 observations
Mine type 5: 65 observations
The classes are around evenly distributed, there is not a major imbalance in mine_type.
Features: the mean voltage is about 0.431, the mean height is about 0.509, and the mean soil value is about 0.504.

#Question 2.2

from sklearn.model_selection import train_test_split
X = data[["voltage", "height", "soil"]]
y = data["mine_type"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.5, random_state=42, stratify=y)

The data is split into 50% test data and 50% training data.
Stratification is used so that each type of mine is represented in both datasets

#Question 2.3

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

for k in range(1, 11):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    predictions = knn.predict(X_test)
    print(k, accuracy_score(y_test, predictions))
  
  I used a k-nearest neighbors (k-NN) classifier to predict the type of mine based on its voltage, height, and soil measurements.
  k-NN works by looking at the k observations that are closest to a new observation and using their mine types to make a prediction.
  I tested different values of k and calculated the accuracy for each one.
  I then chose the value of k with the highest accuracy because it performed the best at predicting the mine type.
  The average validation accuracy for k=2 was 43.7%, so I used k=2 for the final model.


#Question 2.4

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

knn = KNeighborsClassifier(n_neighbors=2)

knn.fit(X_train, y_train)

predictions = knn.predict(X_test)

print(confusion_matrix(y_test, predictions))

print(accuracy_score(y_test, predictions))

The accuracy is 43.8%, so the model correctly predicted 43.8% of test observations
In the matrix, the diagonal shows the correct predictions.
Each row is a different mine type, and through the matrix, you can see that mine type 2 had the greatest prediction accuracy (32)
The model had a difficult time predicting the accuracy of mine type 5 (0)
Overall, the model was good at predicting mine type 1 and mine type 2, and worse at predicting the rest.

#Question 2.5

I would not recommend using this model by itself to decide whether a land mine is safe to remove. The model is only about 43.8% accurate overall, and it makes many mistakes between mine types 3, 4, and 5.
A wrong classification could be very dangerous. I would use the model as an additional source of information that could help experts identify possible mine types.
The prediction could be combined with other information and checked by trained professionals before a mine is removed.
The model could be useful for narrowing down possible mines to remove, but it should not be treated as the sole reliable answer because of the number of incorrect predictions it makes.

### *AI-proposed changes


[Paste the AI's concise numbered list verbatim.]

1. (for question 1)
Your answers are solid overall. Everything is correct and covers what each question asks. Here's a question-by-question check, plus a few places where you could tighten things up.
Q1: Correct and complete. Clear definitions with good examples.
Q2: Correct. You could add that the table lets you compute metrics beyond accuracy (precision, recall, etc.), and that it reveals imbalances, like a model that looks accurate only because it always predicts the majority class. That's the real payoff of a confusion table, so it's worth a sentence.
Q3: Correct. One small addition: SSE is sensitive to outliers because errors are squared, and it can't be compared across datasets of different sizes (that's why people often use MSE or RMSE). It's optional, but it shows deeper understanding.
Q4: Correct. Strong answer. If you want to be thorough, mention that overfitting means low training error but high test error, and underfitting means high error on both. You already say this in words, so this is just a matter of making the contrast explicit.
Q5: Mostly right, but there's one issue. Your explanation of why a split helps is good. However, the question asks why this improves model performance, and you've tied your answer to "different values of k." That assumes the course context is k-nearest neighbors. If your class has been using KNN, that's fine and probably expected. If the question is meant to be general, rephrase it as "different models or hyperparameters (such as k in KNN)." Also, to be precise: the split doesn't improve the model directly. It gives an honest estimate of performance on new data, which lets you choose a better model and avoid overfitting. Your answer says this, but stating it explicitly would make it airtight. One more point some instructors look for: if you tune on the test set repeatedly, the test set is no longer truly unseen, which is why a separate validation set is sometimes used.
Q6: Correct, but the weakest answer. You do cover both sides, but the strengths and weaknesses of the label approach are a bit thin. Consider adding:
Class label strengths: simple, easy to interpret, directly actionable (you get a decision).
Class label weaknesses: no information about confidence, so a 51% prediction looks the same as a 99% one; it's also hard to adjust for different costs of errors.
Probability strengths: conveys uncertainty, lets you set your own threshold (for example, flag a disease at 30% rather than 50% if missing it is costly), and lets you rank cases by risk.
Probability weaknesses: harder to interpret, still requires a threshold for a final decision, and probabilities can be poorly calibrated (an "80%" prediction may not really be right 80% of the time).
Your answer starts with "Example:", which makes it read like the example is the whole answer. I'd restructure it so the strengths and weaknesses are stated directly, with your patient example as an illustration.
Bottom line: Nothing is wrong, and this would likely earn good marks. The biggest improvements would be expanding Q6 and clarifying the "k" reference in Q5. If you have a rubric or the point values, send it over and I can check against it more precisely.

2. ...

QUESTION 2:
I ran your code on the uploaded `land_mines.csv`, and your numbers match. The shape (338, 4), the class counts, the means, and the confusion matrix all reproduce, and k=2 gives 43.8%. Here's the question-by-question check.

**2.1 EDA: correct but thin.** Fix the typo "the The" in your first comment. The question asks you to summarize the features, and you only give three overall means. Two easy additions:
- Voltage separates the classes much more than the other features. Mean voltage is about 0.72 for type 2, versus about 0.30 to 0.40 for the others. Height and soil have nearly identical means across classes (around 0.5). This also sets up your later finding that type 2 is predicted well.
- All three features look like they're already on a 0 to 1 scale, so you don't need to standardize them for k-NN. That matters because k-NN is distance-based. Also mention there are no missing values.

**2.2 Split: correct.** Good that you used stratify.

**2.3 Choosing k: right idea, but two wording problems.**
- You call k=2's 43.7% the "average validation accuracy." It's one accuracy on one test set, not an average, and the 43.7% here doesn't match the 43.8% in 2.4 (the exact value is 43.79%). Use 43.8% in both places.
- You should state that you chose k using the test/validation set, which the question allows. A sentence acknowledging that the final accuracy may be a bit optimistic because k was picked on the same data would strengthen this. The accuracies for k=1 to 10 range from about 31% to 44%, so the differences are small.

**2.4 Confusion table: correct, but you can say more.** Your points are right (type 2 best, type 5 worst), but compute rates instead of raw counts. From your matrix:
- Type 2: 32 of 35 correct (91%)
- Type 1: 25 of 36 (69%)
- Type 3: 9 of 33 (27%)
- Type 4: 8 of 33 (24%)
- Type 5: 0 of 32 (0%)

Your wording "difficult time predicting the accuracy of mine type 5" is awkward. Say "the model never correctly predicted type 5."

**2.5 Practical use: sensible, but it misses what the question is hinting at.** The prompt says you can have a lot of accurate predictions for a type but still make a lot of mistakes. That's about reading the matrix by column (when the model *says* a type) as well as by row. Look at type 1: it's found 69% of the time, but the model predicts "1" 64 times and only 25 are right, so a "1" prediction is correct only about 39% of the time. Types 3, 4, and 5 get dumped into type 1 a lot (12, 12, and 15 times). A "2" prediction, by contrast, is right about 78% of the time (32 of 41). Your answer says types 3, 4, and 5 get confused, which is true, but the bigger pattern is that type 1 acts as a catch-all.

To address this, I'd add advice like:
- Trust a "type 2" prediction much more than a "type 1," "3," or "4" prediction.
- Treat a predicted type 5 as basically unreliable, since the model almost never predicts it.
- Use the model only to narrow down candidates or prioritize which mines to investigate, and have experts confirm before anyone handles a mine.
- Collect more or better features, since these three don't separate types 3, 4, and 5 well.

One optional point: k=2 is an unusual choice because even values of k can produce ties between neighbors. That may partly explain the heavy lean toward type 1, since sklearn breaks ties in favor of the lowest class label. I haven't verified that, so I'd frame it as a possible explanation, and only if your class has covered ties.


EVALUATION OF AI SUGGESTIONS:

For question 1:
Q1: No changes needed. My original answer was already correct, and the only edit was minor rewording for flow.



Q2: Accepted. The suggestion was to mention that a confusion table lets us calculate metrics like precision and recall, and that it can reveal problems accuracy hides, such as a model that always predicts the majority class. I accepted this because it explains why the table is useful beyond just listing TP, TN, FP, and FN, which is what the question asks.



Q3: Accepted. I added that squaring the errors makes outliers have a bigger effect and that SSE grows with dataset size, so it is best for comparing models on the same data.



Q4: Accepted. I changed the wording to "low error on training data but high error on new data" for overfitting, and "high error on both" for underfitting. My original said this in words, but the new version states the contrast more clearly.



Q5: Partially accepted. I accepted the point that the split doesn't directly improve the model but gives an honest estimate of performance on unseen data, which lets us choose a better model. I also changed "different values of k" to "different models or settings (such as different values of k)" so the answer is more general.



Q6: was mostly accepted. I think this was the weakest part of my original answer because it started with an example and didn't clearly state strengths and weaknesses for both approaches. I accepted the restructure so each approach has its own strengths and weaknesses and kept my disease example.

EVALUATION OF AI SUGGESTIONS QUESTION 2:

2.1 Accepted. The AI said my EDA was correct but thin because I only reported overall means. I accepted adding a check for missing values and per-class feature means (groupby('mine_type').mean()). This showed that voltage differs the most across mine types (type 2 averages about 0.72 versus 0.30 to 0.40 for the others), while height and soil are similar across classes. I also accepted the note that the features are already on a 0 to 1 scale, so I didn't standardize them for k-NN.


2.3 Accepted. I fixed two wording problems. I had called 43.7% the "average validation accuracy," but it's one test set accuracy, and it didn't match the 43.8% in 2.4, so I used 43.8% in both places. I also accepted adding that I chose k using the same test/validation set, and that the accuracies for different k were close together.


2.4 Accepted. I replaced raw counts with per-class rates, such as 32 of 35 (91%) for type 2 and 0 of 32 for type 5.


2.5 Accepted. This was the biggest improvement. My original answer said the model was unreliable but didn't address the point in the question about accurate predictions for a type that still come with many mistakes. I added that type 1 is found 69% of the time, but a prediction of type 1 is correct only about 39% of the time, while a prediction of type 2 is correct about 78% of the time. I also added advice to trust type 2 predictions more, not to trust type 5 predictions, to use the model only to narrow down possibilities with expert checking, and to collect more features.
I declined the k=2 tie explanation.


| Change | Decision | Reason and verification |
|---|---|---|
| 1 | Accept / modify / reject | ... |
| 2 | Accept / modify / reject | ... |

### *AI-assisted solution in full
After the evaluation of AI suggestions, incorporate the accepted revisions into your Phase 1 solution and provide the final Phase 2 solution in full in the following cells.

In [ ]:
# Your Phase 2 solution to the problem goes here.


In [ ]:
# Question 1
#1. Regression is used to predict a numerical value. It can be used to predict someone's house price or blood pressure. Classification is used when we want to predict a category or class, such as whether an email is spam or not spam. In short, regression predicts numbers and classification predicts categories.
#2. A confusion table compares the actual class of an observation to the class that the model predicted. It shows true positives, true negatives, false positives, and false negatives, which helps us understand not only how accurate the model is, but also what types of mistakes it is making. It also lets us calculate other metrics beyond accuracy, such as precision and recall, and it can reveal problems that accuracy alone hides. For example, a model that always predicts the majority class can look accurate while never correctly identifying the minority class.
#3. SSE (Sum of Squared Errors) measures how far the model's predicted values are from the actual values. It adds up the squared differences between the actual and predicted values. A lower SSE means the model's predictions are closer to the actual values, so the model fits the data better. Because the errors are squared, large errors have an especially big effect on SSE. SSE also grows with the size of the dataset, so it is best used to compare models on the same data.
#4. Overfitting happens when a model is too complex and learns the training data too closely, including random noise. It has low error on the training data but high error on new data. Underfitting happens when a model is too simple and does not capture the important patterns in the data. An underfit model has high error on both the training and testing data.
#5. Splitting the data allows us to train the model on one set of data and then test it on data the model has not seen before. This helps us see how well the model will perform on new data instead of just seeing how well it memorizes the training data. The split does not directly improve the model, but it gives an honest estimate of performance on unseen data, which lets us choose a better model. By comparing accuracy or SSE for different models or settings (such as different values of k) on the test set, we can choose the one that performs best and avoid overfitting. If we kept tuning on the test set over and over, it would stop being unseen, which is why the test set should be used sometimes.
#6. A class label prediction tells us the single class the model chooses. For example, the model could predict that a patient has a disease or does not have a disease. Its strengths are that it is simple, easy to interpret, and directly actionable because it gives a clear decision. Its weakness is that it does not tell us how confident the model is, so a prediction made with 51% certainty looks the same as one made with 99% certainty. It also gives us no way to adjust for the different costs of errors.
#   A probability distribution gives the probability of each possible class. For example, the model could say there is an 80% chance of disease and a 20% chance of no disease. Its strengths are that it shows the model's confidence, lets us choose our own threshold (for example, flagging a patient at 30% if missing a disease is very costly), and lets us rank cases by risk. Its weaknesses are that it can be more complicated to interpret, we may still need to choose a class based on the probabilities to make a final decision, and the probabilities may not be well calibrated (an "80%" prediction may not actually be correct 80% of the time).


In [1]:
#Question 2.1
import pandas as pd
data = pd.read_csv('land_mines.csv')
print(data.head())
print(data.shape)
print(data.describe())
print(data.isna().sum())                       # new line
print(data['mine_type'].value_counts().sort_index())
print(data.groupby('mine_type').mean())        # new line

# The dataset has 338 observations and 4 variables, with no missing values.
# mine_type is the target variable
# voltage, height, and soil are features used to predict the type of mine
# There are 5 different mine types
# Mine type 1: 71 observations
# Mine type 2: 70 observations
# Mine type 3: 66 observations
# Mine type 4: 66 observations
# Mine type 5: 65 observations
# The classes are roughly evenly distributed, so there is no major imbalance in mine_type.
# Features: the mean voltage is about 0.431, the mean height is about 0.509, and the mean soil value is about 0.504.
# All three features are already on a similar 0 to 1 scale, so I did not standardize them for k-NN.
# Looking at the means by mine type, voltage differs the most across classes. Mine type 2 has a much higher average
# voltage (about 0.72) than the other types (about 0.30 to 0.40). Height and soil have very similar averages
# across all mine types (around 0.5), so they seem to separate the classes less.

    voltage    height  soil  mine_type
0  0.338157  0.000000   0.0          1
1  0.320241  0.181818   0.0          1
2  0.287009  0.272727   0.0          1
3  0.256284  0.454545   0.0          1
4  0.262840  0.545455   0.0          1
(338, 4)
          voltage      height        soil   mine_type
count  338.000000  338.000000  338.000000  338.000000
mean     0.430634    0.508876    0.503550    2.952663
std      0.195819    0.306043    0.344244    1.419703
min      0.197734    0.000000    0.000000    1.000000
25%      0.309737    0.272727    0.200000    2.000000
50%      0.359516    0.545455    0.600000    3.000000
75%      0.482628    0.727273    0.800000    4.000000
max      0.999999    1.000000    1.000000    5.000000
voltage      0
height       0
soil         0
mine_type    0
dtype: int64
mine_type
1    71
2    70
3    66
4    66
5    65
Name: count, dtype: int64
            voltage    height      soil
mine_type                              
1          0.296463  0.495519  0.492958
2 

In [2]:
#Question 2.2

from sklearn.model_selection import train_test_split
X = data[["voltage", "height", "soil"]]
y = data["mine_type"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.5, random_state=42, stratify=y)

# The data is split into 50% training data and 50% test data.
# Stratification is used so that each type of mine is represented in both datasets.

In [3]:
#Question 2.3

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

for k in range(1, 11):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    predictions = knn.predict(X_test)
    print(k, accuracy_score(y_test, predictions))

# I used a k-nearest neighbors classifier to predict the type of mine based on its voltage, height, and soil measurements.
# kNN works by looking at the k observations that are closest to a new observation and using their mine types to make a prediction.
# I tested values of k from 1 to 10 and calculated the test set accuracy for each one.
# I then chose the value of k with the highest accuracy because it performed the best at predicting the mine type.
# k = 2 had the highest accuracy (43.8%), so I used k = 2 for the final model.
# Since k was chosen using the same test/validation set, the final accuracy may be slightly optimistic.
# The accuracies for different values of k were also fairly close to each other, so the choice of k is not very precise.

1 0.40236686390532544
2 0.4378698224852071
3 0.378698224852071
4 0.35502958579881655
5 0.3136094674556213
6 0.3254437869822485
7 0.3136094674556213
8 0.3136094674556213
9 0.34911242603550297
10 0.33727810650887574


In [4]:
#Question 2.4

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

knn = KNeighborsClassifier(n_neighbors=2)

knn.fit(X_train, y_train)

predictions = knn.predict(X_test)

print(confusion_matrix(y_test, predictions))

print(accuracy_score(y_test, predictions))

# The accuracy is 43.8%, so the model correctly predicted 43.8% of test observations.
# In the confusion matrix, each row is the actual mine type and each column is the predicted mine type,
# so the diagonal shows the correct predictions.
# Performance was best for mine type 2, where 32 of 35 were correct (about 91%).
# Mine type 1 was also fairly good, with 25 of 36 correct (about 69%).
# Performance was much worse for the other types: mine type 3 was correct 9 of 33 times (about 27%),
# mine type 4 was correct 8 of 33 times (about 24%), and mine type 5 was never predicted correctly (0 of 32).
# Many mine types 3, 4, and 5 were wrongly predicted as mine type 1.
# Overall, the model was good at predicting mine types 1 and 2, and much worse at the rest.

[[25  0  6  4  1]
 [ 0 32  0  3  0]
 [12  2  9  6  4]
 [12  5  8  8  0]
 [15  2 10  5  0]]
0.4378698224852071


In [5]:
# Question 2.5

# I would not recommend using this model by itself to decide whether a land mine is safe to remove.
# The model is only about 43.8% accurate overall, and a wrong classification could be very dangerous.
# The model can also make many mistakes even for a type it often gets right. For example, mine type 1 is found
# correctly about 69% of the time, but the model predicts type 1 a total of 64 times and only 25 of those are correct
# (about 39%), because many type 3, 4, and 5 mines get labeled as type 1.
# In contrast, when the model predicts type 2, it is correct about 78% of the time (32 of 41).
# So a prediction of type 2 can be trusted much more than a prediction of type 1, 3, or 4, and a prediction
# of type 5 should not be trusted at all since the model almost never predicts it correctly.
# I would use the model only as an additional source of information to help narrow down the possible mine types,
# and every prediction should be combined with other information and checked by trained professionals before a mine is removed.
# To make the model more reliable in the future, more data or additional features would be needed to better
# separate mine types 3, 4, and 5.

### *Reflection on AI assistance
In your own words without generative AI.

[In 150–300 words, describe the main improvements, AI mistakes or limitations, how the final solution was verified, and any remaining concerns.]

I used Claude to check my answers and improve my written explanations. The main improvements were in the analysis. Claude suggested adding per class feature means to the EDA, which showed that voltage separates mine type 2 from the others. It also had me report per class accuracy instead of raw counts, such as 91% for type 2 and 0% for type 5. The biggest improvement was in 2.5, where my original answer said the model was unreliable, but Claude pointed out that a type 1 prediction is correct only about 39% of the time even though 69% of true type 1 mines are found, because many type 3, 4, and 5 mines get labeled as type 1.

But I do think that Claude also made mistakes and had limits. Its first response was cut off, it couldn't see my output until I uploaded the data, and one revision left out the code and showed only comments. It also suggested that k=2 ties might explain the bias toward type 1, which it said it hadn't verified.

I verified the final solution by having the code rerun on the uploaded csv file. My original numbers matched, and I checked the new percentages against the confusion matrix.